### PR TITLE
[SofaBaseLinearSolver] Cleaning in FullMatrix & FullVector & MechanicalState

### PR DESCRIPTION
**Pull request by damienmarchal**
Explicit instanciation of FullMatrix and FullVector for float, double and bool
Move code definition from header (.h) to translation unit (.cpp or .inl).
Forward declare Mat<> in sofa/type/fwd.h
Replace AClass::Index with the generic sofa::Index; 

If you were using AClass::Index, use sofa::Index instead
If you do your own template instanciation don't forget to include the corresponding .inl file instead of the .h

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests][with-regression-tests]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests][with-regression-tests]

File: SofaKernel/modules/Sofa.Type/src/sofa/type/fwd.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+typedef Mat<2,2,SReal> Mat2x2;
+typedef Mat<3,3,SReal> Mat3x3;
+typedef Mat<4,4,SReal> Mat4x4;
+
+typedef Mat<2,2,SReal> Matrix2;
+typedef Mat<3,3,SReal> Matrix3;
+typedef Mat<4,4,SReal> Matrix4;
+}
+
+/// Deprecated namespace.
```
Comment (fredroy) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/fwd.h]: put back those typedefs to compat/sofa/defaulttype/Mat.h

**Comment by damienmarchal** [comment]
Back to review,
Poke @fredroy, @hugtalbot

**Comment by fredroy** [comment]
good for me 😗

**Comment by damienmarchal** [comment]
I removed the conflict... let's go for a CI ride.

**Pull request merged**